File: LLM_experiments/python/results_deepseek-reasoner/solution_try_1/src/cluster_extractor.py
class ClusterExtractor:
    @staticmethod
    def extract_clusters(node, threshold: int):
        if node.name is not None: return [[node.name]]
        if node.height < threshold:
            left = ClusterExtractor.extract_clusters(node.left, threshold)
            right = ClusterExtractor.extract_clusters(node.right, threshold)
            return left + right
        
        left = ClusterExtractor.extract_clusters(node.left, threshold)
        right = ClusterExtractor.extract_clusters(node.right, threshold)
        return [left[0] + right[0]] if left and right else []

File: LLM_experiments/java/results_gpt-4.1/solution_try_2/src/main/java/com/example/phylo/ClusterExtractor.java
package com.example.phylo;

import java.util.ArrayList;
import java.util.List;

public class ClusterExtractor {

    /**
     * Extracts clusters from the tree for a given threshold.
     * Each cluster is a list of species names (leaf nodes).
     */
    public static List<List<String>> extractClusters(PhyloNode root, int threshold) {
        List<List<String>> clusters = new ArrayList<>();
        extractClustersRecursive(root, threshold, clusters);
        return clusters;
    }

    private static void extractClustersRecursive(PhyloNode node, int threshold, List<List<String>> clusters) {
        if (node.isLeaf() || node.getHeight() <= threshold) {
            clusters.add(node.getLeafNames());
        } else {
            extractClustersRecursive(node.getLeft(), threshold, clusters);
            extractClustersRecursive(node.getRight(), threshold, clusters);
        }
    }
}